File: mobile/android/app/src/main/java/com/playbridge/sender/update/UpdateGate.kt
package com.playbridge.sender.update

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import android.widget.Toast
import androidx.compose.runtime.collectAsState

/**
 * Renders the update flow driven by [UpdateChecker.state]. Drop it once near the top of
 * the activity's composition; it shows nothing while [UpdateState.Idle].
 *
 * - [UpdateState.Available] → notify dialog (Play Store or sideload action)
 * - [UpdateState.Downloading] / [UpdateState.Installing] → progress dialog
 * - manual [UpdateState.Checking] / [UpdateState.UpToDate] / [UpdateState.Error] → toast
 */
@Composable
fun UpdateGate(checker: UpdateChecker) {
    val context = LocalContext.current
    val state by checker.state.collectAsState()

    // Lightweight feedback for the manual "Check for updates" path.
    LaunchedEffect(state) {
        when (val s = state) {
            is UpdateState.UpToDate ->
                Toast.makeText(context, "You're on the latest version.", Toast.LENGTH_SHORT).show()
            is UpdateState.Error ->
                if (s.manual) Toast.makeText(context, s.message, Toast.LENGTH_LONG).show()
            else -> Unit
        }
    }

    when (val s = state) {
        is UpdateState.Available -> AvailableDialog(
            info = s.info,
            onAccept = {
                if (s.info.source == InstallSource.SIDELOADED && !checker.canInstall()) {
                    checker.requestInstallPermission()
                } else {
                    checker.accept(s.info)
                }
            },
            onDismiss = checker::dismiss,
        )

        is UpdateState.Downloading -> ProgressDialog(
            title = "Downloading ${s.info.version}",
            fraction = s.fraction,
        )

        is UpdateState.Installing -> ProgressDialog(
            title = "Installing ${s.info.version}",
            fraction = null,
        )

        else -> Unit
    }
}

@Composable
private fun AvailableDialog(
    info: UpdateInfo,
    onAccept: () -> Unit,
    onDismiss: () -> Unit,
) {
    val isPlay = info.source == InstallSource.PLAY_STORE
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Update available") },
        text = {
            Text(
                if (isPlay) {
                    "Version ${info.version} is available. Open the Play Store to update."
                } else {
                    "Version ${info.version} is available. Download and install it now?"
                }
            )
        },
        confirmButton = {
            TextButton(onClick = onAccept) {
                Text(if (isPlay) "Open Play Store" else "Update")
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) { Text("Later") }
        },
    )
}

@Composable
private fun ProgressDialog(title: String, fraction: Float?) {
    AlertDialog(
        onDismissRequest = { /* non-cancelable while working */ },
        title = { Text(title) },
        text = {
            Column(modifier = Modifier.fillMaxWidth()) {
                Spacer(Modifier.height(8.dp))
                if (fraction != null) {
                    LinearProgressIndicator(
                        progress = { fraction.coerceIn(0f, 1f) },
                        modifier = Modifier.fillMaxWidth(),
                    )
                    Spacer(Modifier.height(8.dp))
                    Text(
                        "${(fraction.coerceIn(0f, 1f) * 100).toInt()}%",
                        style = MaterialTheme.typography.bodySmall,
                    )
                } else {
                    LinearProgressIndicator(modifier = Modifier.fillMaxWidth())
                }
            }
        },
        confirmButton = {},
    )
}

File: desktop/lib/update/update_gate.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'update_checker.dart';

/// Renders the update flow driven by [UpdateChecker.state]. Desktop analogue
/// of the Android apps' `UpdateGate` composable: drop it once near the top of
/// the root Stack; it renders nothing while idle.
///
/// Deliberately an inline overlay (modal barrier + card) rather than
/// `showDialog` — declarative rendering from the state machine avoids
/// imperative dialog bookkeeping across async state transitions.
///
/// - [UpdateAvailable]   → "Update available" card (Restart & update / Later)
/// - [UpdateDownloading] → progress card (non-dismissable)
/// - [UpdateInstalling]  → indeterminate card; the app exits mid-state
/// - manual [UpdateUpToDate] / [UpdateError] → auto-dismissing banner
class UpdateGate extends StatefulWidget {
  const UpdateGate({super.key, required this.checker});

  final UpdateChecker checker;

  @override
  State<UpdateGate> createState() => _UpdateGateState();
}

class _UpdateGateState extends State<UpdateGate> {
  Timer? _autoDismiss;

  @override
  void initState() {
    super.initState();
    widget.checker.addListener(_onState);
  }

  @override
  void dispose() {
    widget.checker.removeListener(_onState);
    _autoDismiss?.cancel();
    super.dispose();
  }

  void _onState() {
    _autoDismiss?.cancel();
    final s = widget.checker.state;
    // "You're up to date" is a confirmation, not a decision — clear it itself.
    if (s is UpdateUpToDate) {
      _autoDismiss = Timer(const Duration(seconds: 3), widget.checker.dismiss);
    }
    if (mounted) setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return switch (widget.checker.state) {
      // Quiet non-blocking strip — background checks shouldn't interrupt video.
      UpdateAvailable(:final info) => _availableBanner(info),
      UpdateDownloading(:final info, :final fraction) => _barrier(_progressCard(
          'Downloading v${info.version}…',
          fraction: fraction,
        )),
      UpdateInstalling(:final info) => _barrier(_progressCard(
          'Installing v${info.version}… the app will restart.',
          fraction: null,
        )),
      UpdateError(manual: true, :final message) =>
        _banner(message, error: true),
      UpdateUpToDate() => _banner("You're on the latest version."),
      _ => const SizedBox.shrink(),
    };
  }

  /// Persistent bottom strip with one-tap restart (not a modal barrier).
  Widget _availableBanner(UpdateInfo info) => Positioned(
        left: 0,
        right: 0,
        bottom: 48,
        child: Center(
          child: Material(
            color: Colors.transparent,
            child: Container(
              constraints: const BoxConstraints(maxWidth: 560),
              margin: const EdgeInsets.symmetric(horizontal: 16),
              padding: const EdgeInsets.fromLTRB(16, 12, 8, 12),
              decoration: BoxDecoration(
                color: const Color(0xFF1E1E24),
                borderRadius: BorderRadius.circular(12),
                border: Border.all(
                  color: Colors.tealAccent.withValues(alpha: 0.35),
                ),
                boxShadow: [
                  BoxShadow(
                    color: Colors.black.withValues(alpha: 0.4),
                    blurRadius: 16,
                    offset: const Offset(0, 4),
                  ),
                ],
              ),
              child: Row(
                children: [
                  const Icon(Icons.system_update_alt,
                      size: 20, color: Colors.tealAccent),
                  const SizedBox(width: 12),
                  Expanded(
                    child: Text(
                      'v${info.version} is ready',
                      style: const TextStyle(
                        fontSize: 13,
                        fontWeight: FontWeight.w600,
                      ),
                    ),
                  ),
                  TextButton(
                    onPressed: widget.checker.dismiss,
                    child: const Text('Later'),
                  ),
                  const SizedBox(width: 4),
                  FilledButton(
                    onPressed: () => widget.checker.accept(info),
                    child: const Text('Restart & update'),
                  ),
                  IconButton(
                    tooltip: 'Dismiss',
                    icon: const Icon(Icons.close, size: 16),
                    onPressed: widget.checker.dismiss,
                  ),
                ],
              ),
            ),
          ),
        ),
      );

  Widget _barrier(Widget child) => Positioned.fill(
        child: Container(
          color: Colors.black54,
          alignment: Alignment.center,
          child: child,
        ),
      );

  Widget _card({required List<Widget> children}) => Container(
        width: 420,
        padding: const EdgeInsets.all(24),
        decoration: BoxDecoration(
          color: const Color(0xFF1E1E24),
          borderRadius: BorderRadius.circular(16),
          border: Border.all(color: Colors.white.withValues(alpha: 0.08)),
        ),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: children,
        ),
      );

  Widget _progressCard(String title, {required double? fraction}) =>
      _card(children: [
        Text(title,
            style: const TextStyle(fontSize: 15, fontWeight: FontWeight.w600)),
        const SizedBox(height: 16),
        LinearProgressIndicator(
            value: fraction?.clamp(0.0, 1.0).toDouble(), minHeight: 4),
        if (fraction != null) ...[
          const SizedBox(height: 8),
          Text('${(fraction.clamp(0.0, 1.0) * 100).toInt()}%',
              style: const TextStyle(color: Colors.white54, fontSize: 12)),
        ],
      ]);

  Widget _banner(String message, {bool error = false}) => Positioned(
        left: 0,
        right: 0,
        bottom: 48,
        child: Center(
          child: Container(
            constraints: const BoxConstraints(maxWidth: 520),
            padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 12),
            decoration: BoxDecoration(
              color: const Color(0xFF1E1E24),
              borderRadius: BorderRadius.circular(12),
              border: Border.all(
                color: error
                    ? Colors.redAccent.withValues(alpha: 0.4)
                    : Colors.white.withValues(alpha: 0.08),
              ),
            ),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                Flexible(
                  child: Text(message,
                      style: const TextStyle(fontSize: 13),
                      overflow: TextOverflow.ellipsis,
                      maxLines: 4),
                ),
                if (error) ...[
                  const SizedBox(width: 12),
                  TextButton(
                    onPressed: () {
                      widget.checker.openDownloadPage();
                      widget.checker.dismiss();
                    },
                    child: const Text('Download page'),
                  ),
                ],
                const SizedBox(width: 4),
                IconButton(
                  icon: const Icon(Icons.close, size: 16),
                  onPressed: widget.checker.dismiss,
                ),
              ],
            ),
          ),
        ),
      );
}
